lib/widgets/action_sheet.dart

```diff
@@ -141,6 +142,199 @@ class ActionSheetCancelButton extends StatelessWidget {
+  required String topic,
+}) {
+  final narrow = TopicNarrow(channelId, topic);
+  UserTopicUpdateButton button({
+    UserTopicVisibilityPolicy? from,
+    required UserTopicVisibilityPolicy to,
+  }) {
+    return UserTopicUpdateButton(
+      currentVisibilityPolicy: from,
+      newVisibilityPolicy: to,
+      narrow: narrow,
+      pageContext: context);
+  }
+
+  final mute =                 button(to:   UserTopicVisibilityPolicy.muted);
+  final unmute =               button(from: UserTopicVisibilityPolicy.muted,
+                                      to:   UserTopicVisibilityPolicy.none);
+  final unmuteInMutedChannel = button(to:   UserTopicVisibilityPolicy.unmuted);
+  final follow =               button(to:   UserTopicVisibilityPolicy.followed);
+  final unfollow =             button(from: UserTopicVisibilityPolicy.followed,
+                                      to:   UserTopicVisibilityPolicy.none);
+
+  final store = PerAccountStoreWidget.of(context);
+  final channelMuted = store.subscriptions[channelId]?.isMuted;
+  final visibilityPolicy = store.topicVisibilityPolicy(channelId, topic);
+
+  // TODO(server-7): simplify this condition away
+  final supportsUnmutingTopics = store.connection.zulipFeatureLevel! >= 170;
+  // TODO(server-8): simplify this condition away
+  final supportsFollowingTopics = store.connection.zulipFeatureLevel! >= 219;
+
+  final optionButtons = <Widget>[];
+  if (channelMuted != null && !channelMuted) {
+    switch (visibilityPolicy) {
+      case UserTopicVisibilityPolicy.muted:
+        optionButtons.add(unmute);
+        if (supportsFollowingTopics) {
+          optionButtons.add(follow);
+        }
+      case UserTopicVisibilityPolicy.none:
+      case UserTopicVisibilityPolicy.unmuted:
+        optionButtons.add(mute);
+        if (supportsFollowingTopics) {
+          optionButtons.add(follow);
+        }
+      case UserTopicVisibilityPolicy.followed:
+        optionButtons.add(mute);
+        if (supportsFollowingTopics) {
+          optionButtons.add(unfollow);
+        }
+      case UserTopicVisibilityPolicy.unknown:
+        assert(false);
+        return;
+    }
+  } else if (channelMuted != null && channelMuted) {
+    if (supportsUnmutingTopics) {
+      switch (visibilityPolicy) {
+        case UserTopicVisibilityPolicy.none:
+        case UserTopicVisibilityPolicy.muted:
+          optionButtons.add(unmuteInMutedChannel);
+          if (supportsFollowingTopics) {
+            optionButtons.add(follow);
+          }
+        case UserTopicVisibilityPolicy.unmuted:
+          optionButtons.add(mute);
+          if (supportsFollowingTopics) {
+            optionButtons.add(follow);
+          }
+        case UserTopicVisibilityPolicy.followed:
+          optionButtons.add(mute);
+          if (supportsFollowingTopics) {
+            optionButtons.add(unfollow);
+          }
+        case UserTopicVisibilityPolicy.unknown:
+          assert(false);
+          return;
```

Comment:
Yeah. By having this `unknown` value in the enum we're already paying most of the cost of gracefully handling the situation where the server introduces a new policy value, so we should handle it here too.

Reply:
We do not expect it here because we already explicitly keep it out of our data structure, so it would be a bug to get `unknown` here.

https://github.com/zulip/zulip-flutter/blob/9e42f2694afffd3a40c7ca4786e8641622513350/lib/model/channel.dart#L216-L222

Reply:
Hmm, indeed. In that case it doesn't matter what this line does — though it's still probably cleaner to have it skip this type of button rather than abort the whole function. And it should have a comment saying why it's impossible.

I've noticed in a few other places that it'd be good to have a type for "visibility policy, but only the known/valid ones". Probably that points to removing `unknown` from the enum, and using a nullable type where we actually want that value. Out of scope for this PR, though.

Reply:
Yeah, this will allow us to remove the cases [here](https://github.com/search?q=repo%3Azulip%2Fzulip-flutter+case+UserTopicVisibilityPolicy.unknown%3A+assert%28false%29&type=code) (where this new `assert(false)` originally comes from). It is good to have the type confirm that we process the unknown values at the edge.

Reply:
Opened #1074 and include TODO comments linked in it.